### PR TITLE
[Pointcloud] New interfaces and automatic asset reloading

### LINKED ISSUE #40: [Pointcloud] Eliminate blocking loading of pointcloud assets
As @gadams3 pointed in his review, the pointclouds are currently loaded, blocking main thread. It is not o3de way to deal with loading assets. 
Add necessary handlers to components / Cloud Feature Processor.

### PR DESCRIPTION
**Pull request by Fireronin**
Resolves #40

Adds AcquirePointcloudFromAsset and  UpdatePointCloud
Interfaces to feature processor.

First one creates a ponint cloud based on asset and keeps track of it's changes.
Second one allows for partial updates to pointcloud data

### CODE REVIEW

**Comment by Fireronin** [comment]
I'm unable to fix those formatting issues like I mash reformat and nothing happens, I think my formatting style is correct as I was able to fix other issues using the same formatting

**Review by michalpelka — COMMENTED** [review]
Please apply clang-format-15

File: Gems/Pointcloud/Code/Source/Render/PointcloudFeatureProcessor.cpp [hunk 1]
```diff
@@ -111,6 +159,10 @@ namespace Pointcloud
         {
             UpdateDrawPacket();
         }
+        if (auto it = m_pointcloudAssets.find(asset.GetId()); it != m_pointcloudAssets.end())
+        {
+            UpdatePointCloud(it->second.first, it->second.second->m_data, 0);
```
Comment (michalpelka) [inline Gems/Pointcloud/Code/Source/Render/PointcloudFeatureProcessor.cpp:164]: I am no big fan of second.first.third it is spaghetti code.

File: Gems/Pointcloud/Code/Source/Render/PointcloudFeatureProcessor.h [hunk 2]
```diff
@@ -90,5 +95,6 @@ namespace Pointcloud
         AZ::RHI::Ptr<AZ::RHI::ShaderResourceGroupLayout> m_drawSrgLayout; //!< Shader resource group layout for the draw packet
         AZStd::unordered_map<PointcloudHandle, PointcloudData> m_pointcloudData; //!< Map of pointcloud data
         PointcloudHandle m_currentPointcloudDataIndex = 0; //!< Index to the next pointcloud data to be created
+        AZStd::unordered_map<AZ::Data::AssetId, AZStd::pair<PointcloudHandle, AZ::Data::Asset<PointcloudAsset>>> m_pointcloudAssets;
```
Comment (michalpelka) [inline Gems/Pointcloud/Code/Source/Render/PointcloudFeatureProcessor.h:98]: Why AZ::Data::Asset<PointcloudAsset>> is not a field in Pointcloud data?

Comment (Fireronin) [reply Gems/Pointcloud/Code/Source/Render/PointcloudFeatureProcessor.h]: yaa I guess makes sense

File: Gems/Pointcloud/Code/Source/Render/PointcloudFeatureProcessor.cpp (hunk 1 above)
Comment (Fireronin) [reply Gems/Pointcloud/Code/Source/Render/PointcloudFeatureProcessor.cpp]: resolved

**Comment by michalpelka** [comment]
I've fixed code with help of o3de Discord : https://discord.com/channels/805939474655346758/816043601754325023/1271239320698748951
Now hot-realoding works as expected. 
Our error was, that we have old and new asset. We should keep reference to `AZ::Asset<AZ::Data>` and we are responsible to swap this reference. Without that we are stuck with reference to old data. When reference count on old asset goes to zero, it is destroyed. It is smart, but requires understanding. 

[untitled.webm](https://github.com/user-attachments/assets/413e66c7-4c15-4427-9237-ce8d69dfd819)

With permission from @Fireronin, I've forced pushed code.

**Review by michalpelka — APPROVED** [review]

**Comment by michalpelka** [comment]
I've added @norbertprokopiuk @zakmat  to review, since they are using this code.

File: Gems/Pointcloud/Code/Source/3rdParty/happly.h
Comment (zakmat) [inline Gems/Pointcloud/Code/Source/3rdParty/happly.h:1]: 3rdParty code should be excluded from the auto-formatter

Comment (michalpelka) [reply Gems/Pointcloud/Code/Source/3rdParty/happly.h:1]: Awww, Snap!

File: Gems/Pointcloud/Code/Source/Tools/Components/PointcloudEditorComponent.cpp [hunk 3]
```diff
@@ -66,19 +66,14 @@ namespace Pointcloud
...
                     AZ_Assert(m_featureProcessor, "Failed to enable PointcloudFeatureProcessorInterface.");
-                    m_pointcloudAsset.QueueLoad();
-                    m_pointcloudAsset.BlockUntilLoadComplete();
-                    m_numPoints = m_pointcloudAsset->m_data.size();
-                    AZStd::vector<AZStd::vector<PointcloudAsset::CloudVertex>> cloudVertexDataChunks;
-
-                    m_pointcloudHandle = m_featureProcessor->AcquirePointcloud(m_pointcloudAsset->m_data);
-
-                    if (m_pointcloudHandle != PointcloudFeatureProcessorInterface::InvalidPointcloudHandle)
+                    if (m_pointcloudAsset)
```
Comment (zakmat) [inline Gems/Pointcloud/Code/Source/Tools/Components/PointcloudEditorComponent.cpp:74]: visibility should be taken into account during component activation. Currently if the entity is marked as hidden pointcloud is still rendered upon level activation

File: Gems/Pointcloud/Code/Source/Render/PointcloudFeatureProcessor.cpp [hunk 4]
```diff
@@ -60,37 +61,84 @@ namespace Pointcloud
...
+        pcData.m_assetId = pointcloudAsset.GetId();
+        AZ::Data::AssetBus::MultiHandler::BusConnect(pointcloudAsset.GetId());
+        if (!(pointcloudAsset.IsLoading() || pointcloudAsset.IsReady()))
+        {
+            pointcloudAsset.QueueLoad();
+        }
+        if (pointcloudAsset.IsReady())
+        {
+            OnAssetReady(pointcloudAsset);
+        }
```
Comment (zakmat) [inline Gems/Pointcloud/Code/Source/Render/PointcloudFeatureProcessor.cpp:140]: genuine question: what's the benefit of OnAssetReady at this point?

**Review by zakmat — CHANGES_REQUESTED** [review]
This is a great improvement. Thanks! I would love to see such handling in other places where we BlockUntilCompleted

There is a small but significant bug in the component's Activate - no check for visibility.
Additionally, the m_numPoints is no longer updated.

File: Gems/Pointcloud/Code/Source/Tools/Components/PointcloudEditorComponent.cpp [hunk 5]
```diff
@@ -66,19 +66,14 @@ namespace Pointcloud
...
             [this]()
             {
+                m_scene = AZ::RPI::Scene::GetSceneForEntityId(GetEntityId());
                 if (m_scene && m_pointcloudAsset)
                 {
                     m_featureProcessor = m_scene->EnableFeatureProcessor<PointcloudFeatureProcessor>();
                     AZ_Assert(m_featureProcessor, "Failed to enable PointcloudFeatureProcessorInterface.");
-                    m_pointcloudAsset.QueueLoad();
-                    m_pointcloudAsset.BlockUntilLoadComplete();
-                    m_numPoints = m_pointcloudAsset->m_data.size();
```
[the remark this answers is not in the document]
Comment (Fireronin) [reply Gems/Pointcloud/Code/Source/Tools/Components/PointcloudEditorComponent.cpp:75]: This feature will be dropped as it's quite difficult to do with asynchronous loading we might just leave it as print to console.

File: Gems/Pointcloud/Code/Source/Render/PointcloudFeatureProcessor.cpp (hunk 4 above)
Comment (Fireronin) [reply Gems/Pointcloud/Code/Source/Render/PointcloudFeatureProcessor.cpp:140]: I'm not an expert but from what I understand this asset might have been loaded by some other component then OnAssetReady will not fire by itself

**Review by zakmat — APPROVED** [review]
Looks good!

**Pull request merged**